workflow/util/merge.go

```diff
@@ -37,3 +39,52 @@ func MergeTo(patch, target *wfv1.Workflow) error {
 	}
 	return nil
 }
+
+// MergeMap will merge all element from right map to left map if it is not present in left.
+func MergeMap(left, right map[string]string) {
+	for key, val := range right {
+		if _, ok := left[key]; !ok {
+			fmt.Println(key, val)
+			left[key] = val
+		}
+	}
+}
+
+// MergeWfSpece will merge the workflow space with order of precedence.
+// 1. Workflow Spec, 2 WorkflowTemplate Spec (WorkflowTemplateRef), 3. WorkflowDefault Spec.
+func MergeWfSpecs(wfSpec, wftSpec, wfDefaultSpec *wfv1.WorkflowSpec) (*wfv1.Workflow, error) {
+	if wfSpec == nil {
+		return nil, fmt.Errorf("invalid Workflow spec")
+	}
+	targetWf := wfv1.Workflow{Spec: *wfSpec.DeepCopy()}
+	if wftSpec != nil {
+		err := MergeTo(&wfv1.Workflow{Spec: *wftSpec.DeepCopy()}, &targetWf)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if wfDefaultSpec != nil {
+		err := MergeTo(&wfv1.Workflow{Spec: *wfDefaultSpec.DeepCopy()}, &targetWf)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return &targetWf, nil
+}
+
+// MergeMetaData will merge the patch metadata into target metadata.
+// This func will merge only labels and annotations.
+func MergeMetaDataTo(patch, targetMetaData *metav1.ObjectMeta) {
+	if patch != nil && patch.Labels != nil {
```

Comment:
Perhaps we should just have a single

```go
if patch == nil {
  return
}
```

instead of checking this consistently